File: python/speck.py
MAX64 = (1 << 64) - 1
ROUNDS = 34

def rot64(n, rotation):
    if (rotation > 0):
        return ((n >> (64 - rotation)) | (n << rotation)) & MAX64
    else:
        rotation = -rotation
        return ((n << (64 - rotation)) | (n >> rotation)) & MAX64

def add1(n):
    for i in range(len(n)):
        if(n[i] == MAX64):
            n[i] = 0
        else:
            n[i] += 1
            break

def key_schedule(key):
    key_schedule = [key[0]]
    l = [key[1], key[2], key[3]]

    for i in range(ROUNDS - 1):
        l[i % 3] = ((key_schedule[i] + rot64(l[i % 3], -8)) & MAX64) ^ i
        key_i = rot64(key_schedule[i], 3) ^ l[i % 3]
        key_schedule.append(key_i)

    return key_schedule

def enc_round(n, key):
    left = n[1]
    right = n[0]

    n[1] = ((rot64(left, -8) + right) & MAX64) ^ key
    n[0] = rot64(right, 3) ^ ((rot64(left, -8) + right) & MAX64) ^ key

def speck_encrypt(pt, keys):
    ct = [pt[0], pt[1]]

    for i in range(ROUNDS):
        enc_round(ct, keys[i])

    return ct

def dec_round(n, key):
    left = n[1]
    right = n[0]

    n[1] = rot64((((left ^ key) - rot64(left ^ right, -3) % (MAX64 + 1)) + (MAX64 + 1)) & MAX64, 8)
    n[0] = rot64(left ^ right, -3)

def speck_decrypt(ct, keys):
    pt = [ct[0], ct[1]]

    for i in range(ROUNDS):
        dec_round(pt, keys[i])

    return pt

def speck_ctr(pt, key, nonce):
    out = [0] * len(pt)
    local_nonce = [nonce[0], nonce[1]]
    keys = key_schedule(key)

    for i in range(0, len(pt), 2):
        pad = speck_encrypt(local_nonce, keys)
        out[i] = pt[i] ^ pad[0]
        out[i + 1] = pt[i + 1] ^ pad[1]

        add1(local_nonce)

    return out

File: c/speck.c
#include "speck.h"

#define MAX64 -1UL
#define ROUNDS 34

uint64_t rotate(uint64_t in, int8_t rotation)
{
    if (rotation >= 0) // left circular shift
    {
        return (in >> (64 - rotation)) | (in << rotation);
    }
    else // right circular shift
    {
        rotation = -1 * rotation;
        return (in << (64 - rotation)) | (in >> rotation);
    }
}

void add1(uint64_t * in, size_t length)
{
    int i = 0;
    do
    {
        if (in[i] == MAX64)
        {
            in[i] = 0x0UL;
            i++;
        }
        else
        {
            in[i] += 1;
            break;
        }
    } while (i < length);
}

void key_schedule(uint64_t *key, uint64_t *key_schedule)
{
    uint64_t l[3] = {key[1], key[2], key[3]};
    int i;
    key_schedule[0] = key[0];

    for (i = 0; i < ROUNDS - 1; i++)
    {
        l[i % 3] = (key_schedule[i] + rotate(l[i % 3], -8)) ^ (uint64_t)(i);
        key_schedule[i + 1] = rotate(key_schedule[i], 3) ^ l[i % 3];
    }
}

void enc_round(uint64_t *in, uint64_t key)
{
    uint64_t left = in[1];
    uint64_t right = in[0];

    in[1] = (rotate(left, -8) + right) ^ key;
    in[0] = rotate(right, 3) ^ (rotate(left, -8) + right) ^ key;
}

void speck_encrypt(uint64_t *in, uint64_t *out, uint64_t *keys)
{
    out[0] = in[0];
    out[1] = in[1];

    for (int i = 0; i < ROUNDS; i++)
    {
        enc_round(out, keys[i]);
    }
}

//The data has to be 128 bit aligned
void speck_ctr(uint64_t *in, uint64_t *out, size_t pt_length, uint64_t *key, uint64_t *nonce)
{
    uint64_t pad[2] = {0UL, 0UL};
    uint64_t local_nonce[2] = {nonce[0], nonce[1]};
    uint64_t *keys = malloc(ROUNDS * sizeof(uint64_t));

    if(pad == NULL || keys == NULL) return;

    key_schedule(key, keys);

    for (int i = 0; i <= pt_length - 2; i+=2)
    {
        speck_encrypt(local_nonce, pad, keys);

        out[i] = in[i] ^ pad[0];
        out[i + 1] = in[i + 1] ^ pad[1];

        add1(local_nonce, 2);
    }
    free(keys);
}
